odb_extractor_atoms.py

```diff
@@ -59,98 +223,297 @@ def select_obsclass(classes):
+def autocorr_lag(x, plot=False):
+    '''Test for patterns with auto-correlation'''
+
+    # Auto correlation
+    result = np.correlate(x, x, mode='full')
+    corrmax = np.max(result)
+    if corrmax != 0.0:
+        result = result / corrmax
+    #     print(result)
+    #     print(result.size, result.size//2)
+    if plot:
+        plt.plot(result[result.size // 2:])
+
+    # Pattern offset using first prominent peak
+    # lag = np.argmax(result[result.size//2 + 1:]) + 1
+    peaks, prop = find_peaks(result[result.size // 2:], height=(0, None), prominence=(0.25, None))
+    #     print(peaks)
+    #     print(prop)
+    lag = 0
+    if len(peaks) > 0:
+        lag = peaks[0]
+
+    return lag
+
+# --------------
+
+
+def findatoms(observation, verbose=False, ws=None, fid=sys.stdout):
+    """Analyze a json observing sequence from the ODB and define atoms."""
 
     classes = []
+    guiding = []
     qastates = []
     atoms = []
     natom = 0
-    nabba = 0
 
     # Make dictionary out of obslog to get QA state
     obslog = {}
-    for log_entry in observation['obsLog']:
-        obslog[log_entry['label']] = {'qaState': log_entry['qaState'], 'filename': log_entry['filename']}
-    datalabels = list(obslog.keys())
+    datalabels = []
+    if 'obsLog' in observation.keys():
+        for log_entry in observation['obsLog']:
+            obslog[log_entry['label']] = {'qaState': log_entry['qaState'], 'filename': log_entry['filename']}
+        datalabels = list(obslog.keys())
 
     # Sequence analysis
     sequence = observation['sequence']
     nsteps = len(sequence)
 
+    # First pass to analyze offsets and exptimes/coadds
+    do_not_split = False
+    exptimes = []
+    coadds = []
+    qoffsets = []
+    poffsets = []
+    qoffsets_sky = []
+    poffsets_sky = []
+    config = {'inst': '', 'fpu': [], 'disperser': [], 'filter': [], 'wavelength': []}
+    for ii, step in enumerate(sequence):
+        step_keys = list(step.keys())
+
+        inst = step['instrument:instrument']
+        #     print(inst, fpuinst[inst])
+        if inst == 'Visitor Instrument':
+            inst = step['instrument:name'].split(' ')[0]
+            if inst in ["'Alopeke", "Zorro"]:
+                fpu = 'None'
+            else:
+                fpu = inst
+        else:
+            fpu = step[fpuinst[inst]]
+        config['inst'] = inst
+        config['fpu'].append(fpu)
+
+        if 'instrument:disperser' in step_keys:
+            disperser = step['instrument:disperser']
+        elif inst in ['IGRINS', 'MAROON-X']:
+            disperser = inst
+        else:
+            disperser = 'None'
+        if inst == 'GNIRS':
+            if step['instrument:acquisitionMirror'] == 'in' and \
+                    step['instrument:decker'] == 'acquisition':
+                disperser = 'mirror'
+            else:
+                disperser = disperser.strip('grating') + step['instrument:crossDispersed']
```

Comment:
Is the `disperser.strip('grating')` really what we want? Assuming `disperser` is a `str`, then this just removes any leading or trailing characters in:
```
['g', 'r', a', t', 'i', 'n']
```
from `disperser` and returns the result, which doesn't intuitively seem like what we would want to do.

Including the last `g` in `grating` isn't necessary since it treats this as a set of characters to remove from the front and end of the value contained in `disperser`.

Looking at the output for:
http://gnodbscheduler:8442/programexport?id=GN-2018B-Q-101

and seeing the entries in the sequence such as:
```
"instrument:disperser" : "32 l/mm grating"
```
I can see why this implementation does (I suspect) what we want it to do, but I think what we actually want is:
```
disperser.replace('grating', '')
```

Then, if we have:
```
disperser = "32 l/mm grating"
step['instrument:crossDispersed'] = 'SXD'
```
Doing:
```
disperser = disperser.replace('grating', '') + step['instrument:crossDispersed']
```
gives us:
```
disperser = '32 l/mm SXD'
```
which is - I assume - what we want?